File: Foundation/Unity3D/Encoding.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using cn.vimfung.luascriptcore;
using System;
using MiniJSON;


namespace cn.vimfung.luascriptcore.modules.foundation
{
	/// <summary>
	/// 编码类
	/// </summary>
	public class Encoding : object, LuaExportType 
	{
		/// <summary>
		/// URL编码
		/// </summary>
		/// <returns>编码后内容.</returns>
		/// <param name="text">需要编码文本内容.</param>
		public static string urlEncode(string text)
		{
			return WWW.EscapeURL (text, System.Text.Encoding.UTF8);
		}

		/// <summary>
		/// URL解码
		/// </summary>
		/// <returns>解码后内容.</returns>
		/// <param name="text">需要解码文本内容.</param>
		public static string urlDecode(string text)
		{
			return WWW.UnEscapeURL (text, System.Text.Encoding.UTF8);
		}

		/// <summary>
		/// Base64编码
		/// </summary>
		/// <returns>编码后字符串</returns>
		/// <param name="data">需要编码数据</param>
		public static string base64Encode(object data)
		{
			if (data is byte[] || data is string)
			{
				byte[] rawData = null;
				if (data is string)
				{
					rawData = System.Text.Encoding.UTF8.GetBytes (data as string);
				}
				else
				{
					rawData = data as byte[];
				}

				return Convert.ToBase64String (rawData);
			}

			return null;
		}

		/// <summary>
		/// Base64解码
		/// </summary>
		/// <returns>解码后数据.</returns>
		/// <param name="text">需要解码文本.</param>
		public static byte[] base64Decode(string text)
		{
			return Convert.FromBase64String (text);
		}

		/// <summary>
		/// JSON编码
		/// </summary>
		/// <returns>编码后JSON字符串.</returns>
		/// <param name="obj">需要编码的对象.</param>
		public static string jsonEncode(object obj)
		{
			return Json.Serialize (obj);
		}

		/// <summary>
		/// JSON解码
		/// </summary>
		/// <returns>解码后对象.</returns>
		/// <param name="text">需要解码的JSON字符串.</param>
		public static object jsonDecode(string text)
		{
			return Json.Deserialize (text);
		}

		/// <summary>
		/// 十六进制编码
		/// </summary>
		/// <returns>编码后字符串.</returns>
		/// <param name="data">需要编码的数据.</param>
		public static string hexEncode(object data)
		{
			if (data is byte[] || data is string)
			{
				byte[] rawData = null;
				if (data is string)
				{
					rawData = System.Text.Encoding.UTF8.GetBytes (data as string);
				}
				else
				{
					rawData = data as byte[];
				}

				return BitConverter.ToString (rawData).Replace("-", string.Empty).ToLower();
			}


			return null;
		}

		/// <summary>
		/// 十六进制解码
		/// </summary>
		/// <returns>解码后的数据.</returns>
		/// <param name="text">十六进制编码字符串.</param>
		public static byte[] hexDecode(string text)
		{
			byte[] result = new byte[text.Length / 2];
			for (int x = 0; x < result.Length; x++)
			{
				byte i = Convert.ToByte(text.Substring(x * 2, 2), 16);
				result[x] = (byte)i;
			}
			return result;
		}
	}
}

File: Foundation/Android/cn/vimfung/luascriptcore/modules/foundation/Encoding.java
package cn.vimfung.luascriptcore.modules.foundation;

import android.util.Base64;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URLDecoder;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.Dictionary;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import cn.vimfung.luascriptcore.LuaExportType;

/**
 * 编码工具类
 */
public final class Encoding implements LuaExportType
{
    /**
     * 对文本进行URL编码
     * @param text  文本内容
     * @return  编码后的文本内容
     */
    public static String urlEncode(String text)
    {
        try
        {
            return URLEncoder.encode(text, "utf-8");
        }
        catch (UnsupportedEncodingException e)
        {
            return "";
        }
    }

    /**
     * 对文本进行URL解码
     * @param text  文本内容
     * @return  解码后文本内容
     */
    public static String urlDecode(String text)
    {
        try
        {
            return URLDecoder.decode(text, "utf-8");
        }
        catch (UnsupportedEncodingException e)
        {
            return "";
        }
    }

    /**
     * 对数据进行Base64编码
     * @param data 需要编码数据
     * @return 编码后文本
     */
    public static String base64Encode(Object data)
    {
        if (data instanceof byte[] || data instanceof String)
        {
            byte[] rawData = null;
            if (data instanceof String)
            {
                rawData = ((String) data).getBytes();
            }
            else
            {
                rawData = (byte[]) data;
            }

            return Base64.encodeToString(rawData, Base64.DEFAULT);
        }

        return null;
    }

    /**
     * 对文本进行Base64解码
     * @param text 需要解码文本
     * @return 解码后数据
     */
    public static byte[] base64Decode(String text)
    {
        return Base64.decode(text, Base64.DEFAULT);
    }

    /**
     * 将对象进行JSON编码
     * @param data 需要编码对象
     * @return 编码后的JSON字符串
     */
    public static String jsonEncode(Object data)
    {
        if (data instanceof Map)
        {
            JSONObject object = new JSONObject((Map) data);
            return object.toString();
        }
        else if (data instanceof List)
        {
            JSONArray array = new JSONArray((List) data);
            return array.toString();
        }

        return null;
    }

    /**
     * 将JSON字符串进行解码
     * @param text 需要解码的JSON字符串
     * @return 解码后的对象
     */
    public static Object jsonDecode(String text)
    {

        try
        {

            if (text.matches("^\\s*[\\{].*[\\}]\\s*$"))
            {
                JSONObject object = new JSONObject(text);
                return toMap(object);
            }
            else if (text.matches("^\\s*[\\[].*[\\]]\\s*$"))
            {
                JSONArray array = new JSONArray(text);
                return toList(array);
            }
        }
        catch (JSONException e)
        {

        }

        return null;
    }

    /**
     * 十六进制编码
     * @param data 需要转换成十六进制表示的数据
     * @return 编码后字符串
     */
    public static String hexEncode(Object data)
    {
        if (data instanceof byte[] || data instanceof String)
        {
            byte[] rawData = null;
            if (data instanceof String)
            {
                rawData = ((String) data).getBytes();
            }
            else
            {
                rawData = (byte[]) data;
            }

            if (rawData.length > 0)
            {
                final StringBuilder hexString = new StringBuilder(rawData.length * 2);
                for (byte b : rawData)
                {
                    if ((b & 0xFF) < 0x10)
                    {
                        hexString.append("0");
                    }
                    hexString.append(Integer.toHexString(b & 0xFF));
                }

                return hexString.toString().toLowerCase();
            }
        }

        return null;
    }

    /**
     * 十六进制解码
     * @param text 需要还原回数据的十六进制字符串
     * @return 解码后数据
     */
    public static byte[] hexDecode(String text)
    {
        text = text.toLowerCase();
        final byte[] byteArray = new byte[text.length() / 2];
        int k = 0;
        for (int i = 0; i < byteArray.length; i++)
        {
            //因为是16进制，最多只会占用4位，转换成字节需要两个16进制的字符，高位在先
            byte high = (byte) (Character.digit(text.charAt(k), 16) & 0xff);
            byte low = (byte) (Character.digit(text.charAt(k + 1), 16) & 0xff);
            byteArray[i] = (byte) (high << 4 | low);
            k += 2;
        }
        return byteArray;
    }

    private static Map<String, Object> toMap(JSONObject jsonobj)  throws JSONException
    {
        Map<String, Object> map = new HashMap<String, Object>();
        Iterator<String> keys = jsonobj.keys();
        while(keys.hasNext())
        {
            String key = keys.next();
            Object value = jsonobj.get(key);
            if (value instanceof JSONArray)
            {
                value = toList((JSONArray) value);
            }
            else if (value instanceof JSONObject)
            {
                value = toMap((JSONObject) value);
            }
            map.put(key, value);
        }   return map;
    }

    private static List<Object> toList(JSONArray array) throws JSONException
    {
        List<Object> list = new ArrayList<Object>();
        for(int i = 0; i < array.length(); i++)
        {
            Object value = array.get(i);
            if (value instanceof JSONArray)
            {
                value = toList((JSONArray) value);
            }
            else if (value instanceof JSONObject)
            {
                value = toMap((JSONObject) value);
            }
            list.add(value);
        }   return list;
    }
}
